### ISSUE TITLE
Docs: getting started instructions do not work

### ISSUE THREAD

**Issue by chase-moskal** [typescript-eslint/typescript-eslint#4902]
### Suggested Changes

the docs gave broken advice and was not a smooth experience.

- it asks users to `npm run eslint . --ext .js,.jsx,.ts,.tsx` — but that doesn't work, because the docs never instruct users to *add an `eslint` script to their package.json*
- worse, it advises users to install a `.eslintrc.js`:
    ```js
    module.exports = {
      root: true,
      parser: '@typescript-eslint/parser',
      plugins: [
        '@typescript-eslint',
      ],
      extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/recommended',
      ],
    };
    ```
    but in a modern esm project, this simply fails:
    ```
    Oops! Something went wrong! :(

    ESLint: 8.14.0

    Error [ERR_REQUIRE_ESM]: require() of ES Module /work/namaka/.eslintrc.js from /work/namaka/node_modules/@eslint/eslintrc/dist/eslintrc.cjs not supported.
    .eslintrc.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
    Instead rename .eslintrc.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /work/namaka/package.json to treat all .js files as CommonJS (using .mjs for all ES
    modules instead).
    ```
    skimming some tutorial article somewhere, i learned i could use a `.eslintrc` json instead:
    ```json
    {
      "root": true,
      "parser": "@typescript-eslint/parser",
      "plugins": [
        "@typescript-eslint"
      ],
      "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
      ]
    }
    ```

i'm also perturbed that eslint requires not one, *but two* configuration files *(one config is already too many)* with its `.eslintrc` and also `.eslintignore`

### Affected URL(s)

https://typescript-eslint.io/docs/linting/

**Comment by Zamiell** [comment]
For me, I like using a JS file for my ESLint configuration instead of a JSON file, because you can insert comments, which is really handy for documentation purposes.

e.g.

```js
// This is a shared configuration file for ESLint
// https://eslint.org/docs/user-guide/configuring
module.exports = {
  extends: [
    // The linter base is the Airbnb style guide,
    // which is the most popular JavaScript style guide in the world:
    // https://github.com/airbnb/javascript
    // The actual ESLint config is located here:
    // https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules
    // The TypeScript config extends it:
    // https://github.com/iamturns/eslint-config-airbnb-typescript/blob/master/lib/shared.js
    // This includes the "parser" declaration of "@typescript-eslint/parser"
    "airbnb-base",
    "airbnb-typescript/base",

    // We extend the Airbnb rules with the "recommended" and "recommended-requiring-type-checking"
    // rules from the "typescript-eslint" plugin, which is also recommended by Matt Turnbull,
    // the author of "airbnb-typescript/base"
    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/README.md#recommended
    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended.ts
    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended-requiring-type-checking.ts
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
```

When you say "a modern esm project", do you just mean having `"type": "module"` in your package.json file?
You could try renaming the file to ".mjs" or ".cjs" to see if that fixes it.

> i'm also perturbed that eslint requires not one, but two configuration files (one config is already too many) with its .eslintrc and also .eslintignore

.eslintignore isn't required in most projects, as far as I know.
As far as VSCode goes, ESLint will simply lint whatever is included in your "tsconfig.json" file, so there's no need to do any additional steps.

**Comment by bradzacher** [comment]
> it asks users to npm run eslint . --ext .js,.jsx,.ts,.tsx — but that doesn't work, because the docs never instruct users to add an eslint script to their package.json

NPM changes over time and commands change. Most users should know npm well enough to know how to run a module binary. We don't really use npm that much, so it's not something we keep up with.

The important bit is the flags which instruct you on how to correctly call ESLint from the CLI.

Happy to accept a PR if you want to update this.

> worse, it advises users to install a .eslintrc.js:

`.eslintrc.js` is *the* standard. Using a JSON file is strongly, ***STRONGLY*** discouraged. In future versions of ESLint it straight up won't be supported.
Additionally if you don't use a JS config - you won't be able to properly configure type-aware linting (which is mentioned in the relevant docs).

> but in a modern esm project

Our getting started guide isn't a comprehensive "this is how you can configure every scenario" guide. It's a guide for the most common usecase which provides a foundation for those that have different setups to learn from and tweak to their own world. For example - we don't once mention how to setup webpack - even though it's not uncommon people run eslint via webpack.

Most projects aren't using ESM because it's still relatively bleeding-edge stuff. As Zamiel mentions as well - most (if not all) TS projects aren't using ESM either because it's not currently supported by TS.
If you're setting `type: module` in your package.json - really you should only be doing so if you understand the consequences of such an action - which is that `.js` files won't work and you instead need to use a `.cjs` file.

Finally we also purposely don't go in-depth and explain advanced concepts or anything. For example - it's rather common that people use overrides configs to setup their projects and lint certain files with different configs.
To help people figure out advanced concepts, or how to expand this foundation to their world - [we also directly link out to relevant ESLint docs](https://typescript-eslint.io/docs/linting/#further-configuration-documentation) which, as an example, [directly mention what to do if you're using ESM](https://eslint.org/docs/user-guide/configuring/configuration-files#:~:text=JavaScript%20(ESM)%20%2D%20use%20.eslintrc.cjs%20when%20running%20ESLint%20in%20JavaScript%20packages%20that%20specify%20%22type%22%3A%22module%22%20in%20their%20package.json.%20Note%20that%20ESLint%20does%20not%20support%20ESM%20configuration%20at%20this%20time.).

> i'm also perturbed that eslint requires not one, but two configuration files (one config is already too many)

First off let me address the latter part - "one config is already too many".
How else would you suggest ESLint understands what rules, parser(s) and rule configs you want to use? This config can't be inferred and needs to be explicitly setup.
Regardless though - this feedback isn't relevant to our project - we build on top of ESLint. Feel free to provide them that feedback if you're so motivate and can provide a solution along side the opinion.

Second - you don't *need* an ignore file. We do it because it's easier for most people to understand - it's analogous to a `.gitignore` and shares the same syntax.
Generally we find having a separate ignore file is better because it's the sort of file which is touched once and never again - so having it isolated keeps the config file (which can change more frequently) less cluttered and easier to understand.

Again, however, if you were to delve into the ESLint docs which are linked, you would [find a section on how you can include the ignore config within your ESLint config](https://eslint.org/docs/user-guide/configuring/ignoring-code#ignorepatterns-in-config-files)

---

TL;DR - Happy to accept a PR if you want to:
- improve the run command so it's correct for the npm world
   - it looks like the docasaurus plugin relies on `npm-to-yarn`, which doesn't support run commands - so we'd need to write this as our own tabs.
- add a note to the docs about using a `.cjs` file for `type: module` projects.

**Comment by chase-moskal** [comment]
thanks for providing your thoughts and reasoning, and the historical aspects of the situation.

i just want it to be clear, that the most important issue of this thread is:

- when followed *exactly,* the npm instructions produce failure.
  - i understand the sentiment expressed here:  
      > NPM changes over time and commands change. Most users should know npm well enough to know how to run a module binary. We don't really use npm that much, so it's not something we keep up with.
  - but i still feel it's better for a strict following of the instructions to produce success, instead of producing guaranteed failure that a lowly newcomer has to debug (only to discover that it's the instructions that are wrong, not their reading)

the rest is opinionated, that esm should be a first-class citizen.

as for esm: i understand it may be seen as outside the norm, but for my team, and many teams i've been working with, we have been using es modules exclusively in typescript *for years now.* it's a little bemusing to encounter folks who haven't yet made the upgrade, and to then be treated like the weirdo for merely keeping pace with now-ancient es6 and node 12 ;)

i won't be submitting a pull request — my team wasn't able to configure eslint's indent configuration to enforce tabs while also respecting our function parameter destructuring convention, so we simply had to move on and abandon this linter, in the interest of time.

**Comment by JoshuaKGoldberg** [comment]
A few thoughts here:

> es modules exclusively in typescript _for years now_

We've been using `export` `import` inside `.ts`/`.tsx` source files too, but using `"type": "module"` in `package.json` for _years_? That's a very bold+unusual choice given the lack of wide adoption or platform-level support prior to the last year or so. If that's what you mean, then props to you!

> better for a strict following of the instructions to produce success

Agree. I'll mark this issue explicitly as accepting PRs because the docs should work first-run for anybody using npm.

I also think it'd be a good idea to have a little note somewhere about ESM with a link to the core ESLint docs - as bradzacher posted. Also something(s) in our FAQs summarizing this issue. 

> my team wasn't able to configure eslint's indent configuration to enforce tabs

I'd love to hear more about this. Both what if anything you're using instead of ESLint, and the decision to avoid Prettier (#4907). Feel free to email me at the address mentioned on my GitHub README if you want any help with that!

**Comment by Josh-Cena** [comment]
FWIW, I've also been using `"type": "module"` in my package.json, combined with `"module": "esnext"` in tsconfig, since there's a general trend in the npm ecosystem to migrate to ESM (~two years?). The tooling is not fully ready yet, but ever since Node v14, it's not been broken, and `"module": "esnext"` has always been working fine.

`.eslintrc.js` is certainly getting increasingly intractable to use. I strongly suggest the docs suggest `.eslintrc.cjs` instead to disambiguate.
